bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java

```diff
@@ -236,4 +239,22 @@ private void cleanPropertyBuilders(final ClassModelBuilder<?> classModelBuilder)
             classModelBuilder.removeProperty(propertyName);
         }
     }
+
+    private <T> void processBsonExtraElementsAnnotation(final PropertyModelBuilder<T> propertyModelBuilder) {
+        PropertyAccessor<T> propertyAccessor = propertyModelBuilder.getPropertyAccessor();
+        if (!(propertyAccessor instanceof PropertyAccessorImpl)) {
+            throw new CodecConfigurationException(format("The @BsonExtraElements annotation is not compatible with "
+                            + "propertyModelBuilder instances that have custom implementations of org.bson.codecs.pojo.PropertyAccessor: %s",
+                    propertyModelBuilder.getPropertyAccessor().getClass().getName()));
+        }
+
+        if (!Map.class.isAssignableFrom(propertyModelBuilder.getTypeData().getType())) {
```

Comment:
Only if the type has generic types. So Document would fail here as it has no generic types.

Reply:
I think it could be done.  For example, this:

```
        Type[] genericInterfaces = Document.class.getGenericInterfaces();
        for (var type : genericInterfaces) {
            if (type instanceof ParameterizedType parameterizedType) {
                System.out.println(parameterizedType.getRawType().getTypeName());
                for (var typeArguments : parameterizedType.getActualTypeArguments()) {
                    System.out.println("   " + typeArguments.getTypeName());
                }
            }
        }
```

prints:

```
java.util.Map
   java.lang.String
   java.lang.Object
```

Reply:
Added JAVA-4584 to track that.